libcaf_net/caf/net/stream_transport.hpp

```diff
@@ -198,49 +201,97 @@ class stream_transport {
-    if (write_buf_.empty())
+  void write_packet(unit_t, span<buffer_type*> buffers) {
+    // Sanity check
+    CAF_ASSERT(!buffers.empty());
+    auto it = buffers.begin();
+    if (write_queue_.empty())
       manager().register_writing();
-    write_buf_.insert(write_buf_.end(), header.begin(), header.end());
-    write_buf_.insert(write_buf_.end(), payload.begin(), payload.end());
+    // move header by itself to keep things sorted.
+    write_queue_.emplace_back(true, std::move(**it++));
+    // payload buffers. just write them
+    for (; it != buffers.end(); ++it)
+      write_queue_.emplace_back(false, std::move(**it));
+  }
+
+  // -- buffer recycling -------------------------------------------------------
+
+  buffer_type next_header_buffer() {
+    return next_buffer_impl(free_header_bufs_);
+  }
+
+  buffer_type next_buffer() {
+    return next_buffer_impl(free_bufs_);
+  }
+
+  buffer_type next_buffer_impl(std::deque<buffer_type>& container) {
+    if (container.empty()) {
+      return {};
+    } else {
+      auto buf = std::move(container.front());
+      container.pop_front();
+      return buf;
+    }
   }
 
 private:
   // -- private member functions -----------------------------------------------
 
   bool write_some() {
-    if (write_buf_.empty())
+    auto begin = [&]() { return write_queue_.begin(); };
+    // helper to sort empty buffers back into the right queues
+    auto recycle = [&]() {
+      auto is_header = [](std::pair<bool, buffer_type>& p) { return p.first; };
+      begin()->second.clear();
+      if (is_header(*begin()))
+        free_header_bufs_.emplace_back(std::move(begin()->second));
+      else
+        free_bufs_.emplace_back(std::move(begin()->second));
+      write_queue_.pop_front();
+    };
+    // nothing to write
+    if (write_queue_.empty())
       return false;
-    auto len = write_buf_.size() - written_;
-    auto buf = write_buf_.data() + written_;
-    CAF_LOG_TRACE(CAF_ARG(handle_.id) << CAF_ARG(len));
-    auto ret = write(handle_, make_span(buf, len));
-    if (auto num_bytes = get_if<size_t>(&ret)) {
-      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
-      // Update state.
-      written_ += *num_bytes;
-      if (written_ >= write_buf_.size()) {
-        written_ = 0;
-        write_buf_.clear();
-        return false;
+    do {
+      if (begin()->second.empty()) {
+        recycle();
+        continue;
       }
-    } else {
-      auto err = get<sec>(ret);
-      if (err != sec::unavailable_or_would_block) {
-        CAF_LOG_DEBUG("send failed" << CAF_ARG(err));
-        worker_.handle_error(err);
+      // get size of send buffer
+      auto ret = send_buffer_size(handle_);
+      if (!ret) {
+        CAF_LOG_ERROR("send_buffer_size returned an error" << CAF_ARG(ret));
         return false;
       }
-    }
-    return true;
+      // is send buffer of socket full?
+      if (begin()->second.size() > *ret)
```

Comment:
```suggestion
      if (begin()->second.size() > *ret)
```

I think this check is wrong for two reasons:
- `ret` is the result of `getsockopt` with `SO_SNDBUF`, which returns the *maximum* socket send buffer size and *not* how much space is currently available
- What happens here if a payload is bigger than the socket buffer? That would block forever, no?

I don't think we need to keep track of `send_buffer_size` at all, actually. Rather than checking whether the entire buffer fits into the socket buffer, we should just write and deal with partial writes by storing where to continue writing the front buffer on the next invocation.